File: assign4/python/src/driver.py
import os
import time
import sys
from functools import reduce
sys.path.append('../src')
import gameoflife as GameOfLife

def cls():
    os.system('cls' if os.name=='nt' else 'clear')

def get_window(lis):
  xmax = max(lis, key = lambda tupl: tupl[0])[0]
  ymax = max(lis, key = lambda tupl: tupl[1])[1]
  xmin = min(lis, key = lambda tupl: tupl[0])[0]
  ymin = min(lis, key = lambda tupl: tupl[1])[1]
  return {"xmin":xmin, "xmax":xmax, "ymin":ymin, "ymax":ymax}

def print_list(cell_list):
  win = get_window(cell_list)
 
  xrange = (abs(win["xmin"] - win["xmax"]))
  yrange = (abs(win["ymin"] - win["ymax"]))

  xoffset = win["xmin"]
  yoffset = win["ymin"]

  shifted_cell_list =list(
    map(lambda tup: (tup[0] - xoffset, tup[1] - yoffset), cell_list))

  for y in reversed(range(yrange + 1)):
    for x in range(xrange + 1):
      if (x,y) in shifted_cell_list:
        print("x", end = "")
      else:
        print(" ", end = "")
    print()
                          
def initial():
  return [(1, 1), (1, 3), (2, 2), (2, 3), (3, 2)]

def print_cli(current_gen, ignore):
  cls()
  print_list(current_gen)
  time.sleep(1)
  return GameOfLife.next_generation(current_gen)

time.sleep(4)
reduce(print_cli, iter(initial, 1))

File: assign4/haskell/src/driver.hs
import System.Process
import System.Info
import Control.Concurrent
import Data.List as LS
import Data.Map
import Data.Maybe
import GameOfLife (nextGeneration)


getWindow cells = do
    let xmax = fst $ maximumBy (\x y -> fst x `compare` fst y) cells
    let ymax = snd $ maximumBy (\x y -> snd x `compare` snd y) cells
    let xmin = fst $ minimumBy (\x y -> fst x `compare` fst y) cells
    let ymin = snd $ minimumBy (\x y -> snd x `compare` snd y) cells
    fromList [("xmax", xmax), ("ymax", ymax), ("xmin", xmin), ("ymin", ymin)]

cellsToString cells = do
    let win = getWindow cells

    let xmin = win ! "xmin"
    let xmax = win ! "xmax"
    let ymin = win ! "ymin"
    let ymax = win ! "ymax"

    let xrange = abs $ xmin - xmax
    let yrange = abs $ ymin - ymax

    let xoffset = win ! "xmin"
    let yoffset = win ! "ymin"

    let shiftedCells = LS.map (\cell -> (fst cell - xoffset, snd cell - yoffset)) cells

    let indices = [(x, y) | x <- [0..xrange],
                            y <- [0..yrange]]

    LS.intercalate "" $ LS.map (\cell -> if elem cell shiftedCells then (if snd cell == yrange then "x\n" else "x") else (if snd cell == yrange then " \n" else " ")) indices

printIt cellsString = do
    threadDelay 1000000
    system (if System.Info.os == "windows" then "cls" else "clear")
    putStr cellsString


main = do
    let cells = [(1, 1), (1, 3), (2, 2), (2, 3), (3, 2)]
    let results = iterate (nextGeneration) cells
    threadDelay 2000000
    mapM_ printIt (LS.map cellsToString (LS.take 1000000 results))